pylzy/lzy/env/container/base.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from enum import Enum
+from typing import Optional, ClassVar, TYPE_CHECKING
+from typing_extensions import Protocol, runtime_checkable
+
+from lzy.env.base import Deconstructible
+
+if TYPE_CHECKING:
+    from lzy.env.mixin import WithEnvironmentType
+
+
+class DockerPullPolicy(Enum):
+    ALWAYS = "ALWAYS"  # Always pull the newest version of image
+    IF_NOT_EXISTS = "IF_NOT_EXISTS"  # Pull image once and cache it for next executions
+
+
+@runtime_checkable
+class DockerProtocol(Protocol):
```

Comment:
Why protocol for this?
We can use smth like this:
```python
@final
@dataclass
class Docker:
    ...

class NoContainer:
    pass

Container = Union[Docker, NoContainer]  # can be extended

class ContainerProvider(ABC):

    @staticmethod
    def provide(self) -> Container:
        pass

```

Reply:
We had a conversation in telegram, so wrap up is:
Maybe I could agree with you about protocols is too much, but I don't like your specific proposal, because I want to every Container be an subclass of BaseContainer and be consistent in this way with other parts of environments.

Reply:
Vladimir, let's make it as simple as possible
We have only a single implementation of a container provider, let's drop this interface